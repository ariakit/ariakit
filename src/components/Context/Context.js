import React from 'react'

const Context = React.createContext({})

export default Context
