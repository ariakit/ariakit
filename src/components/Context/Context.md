```jsx
const { Context, Hidden, Button, ContextLogger } = require('reas');

const MyButton = () => (
  <Hidden.State context="foo">
    {hidden => <Button as={Hidden.Toggle} {...hidden}>Toggle</Button>}
  </Hidden.State>
);

const MyHidden = () => (
  <Hidden.State context="foo">
    {hidden => <Hidden {...hidden}>Hidden</Hidden>}
  </Hidden.State>
);

const App = () => (
  <Context.Provider logger={ContextLogger}>
    <MyButton />
    <MyHidden />
  </Context.Provider>
);

<App />
```