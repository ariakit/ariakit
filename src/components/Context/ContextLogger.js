/* eslint-disable no-console */
import { diffString } from 'json-diff'

const logColor = string => {
  const colorMap = {
    '+': 'green',
    '-': 'red',
  }

  string
    .split('\n')
    .filter(x => !!x)
    .forEach(line => {
      const color = colorMap[line.charAt(0)]
      if (color) {
        console.log(`%c${line}`, `color: ${color}`)
      } else {
        console.log(line)
      }
    })
}

const ContextLogger = (prevState, nextState) => {
  const diff = diffString(prevState, nextState, { color: false }).trim()
  if (diff.split('\n').length > 6) {
    console.groupCollapsed('Context Update')
  } else {
    console.group('Context Update')
  }
  console.log('State', nextState)
  logColor(diff)
  console.groupEnd()
}

export default ContextLogger
