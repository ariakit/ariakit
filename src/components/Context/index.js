import Provider from './ContextProvider'
import Consumer from './ContextConsumer'

const Context = {
  Provider,
  Consumer,
}

export default Context
