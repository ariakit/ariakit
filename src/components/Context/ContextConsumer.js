import React from 'react'
import PropTypes from 'prop-types'
import pick from 'lodash/pick'
import Context from './Context'
import mapStateToActions from '../../utils/mapStateToActions'
import mapStateToSelectors from '../../utils/mapStateToSelectors'

const ContextConsumer = ({
  context,
  stateKeys,
  actions,
  selectors,
  children,
  ...props
}) => {
  const parseState = globalState => ({
    ...pick(props, stateKeys),
    ...globalState[context],
  })

  const parseActions = globalSetState =>
    mapStateToActions(
      (fn, cb) =>
        globalSetState(
          globalState => ({
            [context]: {
              ...parseState(globalState),
              ...fn(parseState(globalState)),
            },
          }),
          cb,
        ),
      actions,
    )

  const parseSelectors = globalState =>
    mapStateToSelectors(parseState(globalState), selectors)

  return (
    <Context.Consumer>
      {global =>
        children({
          ...parseState(global.state),
          ...(actions && parseActions(global.setState)),
          ...(selectors && parseSelectors(global.state)),
        })
      }
    </Context.Consumer>
  )
}

ContextConsumer.propTypes = {
  context: PropTypes.string,
  stateKeys: PropTypes.arrayOf(PropTypes.string),
  actions: PropTypes.objectOf(PropTypes.func),
  selectors: PropTypes.objectOf(PropTypes.func),
  children: PropTypes.func.isRequired,
}

export default ContextConsumer
