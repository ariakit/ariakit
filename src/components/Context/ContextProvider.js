import React from 'react'
import PropTypes from 'prop-types'
import Context from './Context'

class ContextProvider extends React.Component {
  static propTypes = {
    children: PropTypes.node,
    logger: PropTypes.func,
  }

  state = {}

  onSetState = fn => {
    if (typeof this.props.logger === 'function') {
      const prevState = { ...this.state }
      this.setState(fn, () => this.props.logger(prevState, this.state))
    } else {
      this.setState(fn)
    }
  }

  render() {
    return (
      <Context.Provider
        value={{
          state: this.state,
          setState: this.onSetState,
        }}
      >
        {this.props.children}
      </Context.Provider>
    )
  }
}

export default ContextProvider
