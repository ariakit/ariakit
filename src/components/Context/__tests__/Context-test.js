import React from 'react'
import { mount } from 'enzyme'
import Context from '..'

const wrap = (props, logger) =>
  mount(
    <Context.Provider logger={logger}>
      <Context.Consumer context="foo" {...props} />
    </Context.Provider>,
  )

test('simple state', () => {
  const initialState = { foo: 'bar' }
  const children = v => <div>{v.foo}</div>
  const wrapper = wrap({ initialState, children })
  expect(wrapper.contains('bar')).toBe(true)
})

test('actions', () => {
  const initialState = { n: 0 }
  const actions = {
    increment: () => s => ({ n: s.n + 1 }),
  }
  const children = v => <button onClick={v.increment}>{v.n}</button>
  const wrapper = wrap({ initialState, actions, children })
  expect(wrapper).toMatchSnapshot()
  wrapper.simulate('click')
  expect(wrapper).toMatchSnapshot()
})

test('selectors', () => {
  const initialState = { foo: 'bar' }
  const selectors = {
    getFoo: () => s => s.foo,
  }
  const children = v => <div>{v.getFoo()}</div>
  const wrapper = wrap({ initialState, selectors, children })
  expect(wrapper.contains('bar')).toBe(true)
})

test('logger', () => {
  const initialState = { n: 0 }
  const logger = jest.fn()
  const actions = {
    increment: () => state => ({ n: state.n + 1 }),
  }
  const children = v => <button onClick={v.increment}>{v.n}</button>
  const wrapper = wrap({ initialState, actions, children }, logger)
  expect(logger).toHaveBeenCalledTimes(1)
  expect(logger).toHaveBeenCalledWith({}, { foo: { n: 0 } })
  wrapper.simulate('click')
  expect(logger).toHaveBeenCalledTimes(2)
  expect(logger).toHaveBeenCalledWith({ foo: { n: 0 } }, { foo: { n: 1 } })
})
