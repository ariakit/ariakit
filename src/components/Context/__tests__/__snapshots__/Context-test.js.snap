// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`actions 1`] = `
<ContextProvider>
  <ContextConsumer
    actions={
      Object {
        "increment": [Function],
      }
    }
    context="foo"
    initialState={
      Object {
        "n": 0,
      }
    }
  >
    <ContextChild
      actions={
        Object {
          "increment": [Function],
        }
      }
      context="foo"
      initialState={
        Object {
          "n": 0,
        }
      }
      setState={[Function]}
      state={
        Object {
          "foo": Object {
            "n": 0,
          },
        }
      }
    >
      <button
        onClick={[Function]}
      >
        0
      </button>
    </ContextChild>
  </ContextConsumer>
</ContextProvider>
`;

exports[`actions 2`] = `
<ContextProvider>
  <ContextConsumer
    actions={
      Object {
        "increment": [Function],
      }
    }
    context="foo"
    initialState={
      Object {
        "n": 0,
      }
    }
  >
    <ContextChild
      actions={
        Object {
          "increment": [Function],
        }
      }
      context="foo"
      initialState={
        Object {
          "n": 0,
        }
      }
      setState={[Function]}
      state={
        Object {
          "foo": Object {
            "n": 1,
          },
        }
      }
    >
      <button
        onClick={[Function]}
      >
        1
      </button>
    </ContextChild>
  </ContextConsumer>
</ContextProvider>
`;
